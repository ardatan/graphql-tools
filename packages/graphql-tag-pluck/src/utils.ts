// Will use the shortest indention as an axis
export const freeText = (text: string | string[], skipIndentation = false) => {
  if (text instanceof Array) {
    text = text.join('');
  }

  // This will allow inline text generation with external functions, same as ctrl+shift+c
  // As long as we surround the inline text with ==>text<==
  text = text.replace(/( *)==>((?:.|\n)*?)<==/g, (_match, baseIndent, content: string) => {
    return content
      .split('\n')
      .map(line => `${baseIndent}${line}`)
      .join('\n');
  });

  if (skipIndentation) {
    if (text.trim() === '') {
      return '';
    }
    return text;
  }

  const lines = text.split('\n');

  const minIndent = lines
    .filter(line => line.trim())
    .reduce<number>((minIndent, line) => {
      const currIndent = line.match(/^ */)?.[0].length;
      if (currIndent == null) {
        return minIndent;
      }

      return currIndent < minIndent ? currIndent : minIndent;
    }, Infinity);

  return lines
    .map(line => line.slice(minIndent))
    .join('\n')
    .trim()
    .replace(/\n +\n/g, '\n\n');
};

// foo_barBaz -> ['foo', 'bar', 'Baz']
export const splitWords = (str: string) => {
  return str.replace(/[A-Z]/, ' $&').split(/[^a-zA-Z0-9]+/);
};

// upper -> Upper
export const toUpperFirst = (str: string) => {
  return str.slice(0, 1).toUpperCase() + str.slice(1).toLowerCase();
};

// foo-bar-baz -> fooBarBaz
export const toCamelCase = (str: string) => {
  const words = splitWords(str);
  const first = words.shift()?.toLowerCase() ?? '';
  const rest = words.map(toUpperFirst);

  return [first, ...rest].join('');
};
