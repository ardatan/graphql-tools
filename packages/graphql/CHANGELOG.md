# @graphql-tools/graphql

## 0.0.1

### Patch Changes

- [#4688](https://github.com/ardatan/graphql-tools/pull/4688) [`72dadfe5`](https://github.com/ardatan/graphql-tools/commit/72dadfe558a18e6ac41da28f5965eb740f4f368e) Thanks [@saihaj](https://github.com/saihaj)! - initial release
