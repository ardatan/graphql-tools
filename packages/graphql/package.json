{
  "name": "@graphql-tools/graphql",
  "version": "0.0.1",
  "author": "Saihajpreet Singh <saihajpreet.singh@gmail.com>",
  "license": "MIT",
  "sideEffects": false,
  "repository": {
    "type": "git",
    "url": "https://github.com/ardatan/graphql-tools.git",
    "directory": "packages/graphql"
  },
  "keywords": [
    "gql",
    "graphql",
    "typescript"
  ],
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "exports": {
    ".": {
      "require": {
        "types": "./dist/typings/index.d.cts",
        "default": "./dist/cjs/index.js"
      },
      "import": {
        "types": "./dist/typings/index.d.ts",
        "default": "./dist/esm/index.js"
      },
      "default": {
        "types": "./dist/typings/index.d.ts",
        "default": "./dist/esm/index.js"
      }
    },
    "./*": {
      "require": {
        "types": "./dist/typings/*.d.cts",
        "default": "./dist/cjs/*.js"
      },
      "import": {
        "types": "./dist/typings/*.d.ts",
        "default": "./dist/esm/*.js"
      },
      "default": {
        "types": "./dist/typings/*.d.ts",
        "default": "./dist/esm/*.js"
      }
    },
    "./package.json": "./package.json"
  },
  "typings": "dist/typings/index.d.ts",
  "typescript": {
    "definition": "dist/typings/index.d.ts"
  },
  "devDependencies": {
    "typescript": "4.8.4"
  },
  "buildOptions": {
    "input": "./src/index.ts"
  },
  "publishConfig": {
    "directory": "dist",
    "access": "public"
  },
  "type": "module"
}
