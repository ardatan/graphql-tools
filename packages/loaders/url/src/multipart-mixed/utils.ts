/* eslint-disable */

export function getBoundaryFromContentType(contentType: string) {
  const idxBoundary = contentType.indexOf('boundary=');
  return `--${
    !!~idxBoundary
      ? // +9 for 'boundary='.length
        contentType
          .substring(idxBoundary + 9)
          .trim()
          .replace(/['"]/g, '')
      : '-'
  }`;
}

export const separator = '\r\n\r\n';

export const decoder = new TextDecoder();
