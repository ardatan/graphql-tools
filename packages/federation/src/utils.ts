import { GraphQLSchema } from 'graphql';
import { FilterRootFields, FilterTypes } from '@graphql-tools/wrap';

export function getArgsFromKeysForFederation(representations: readonly any[]) {
  return { representations };
}

export function getKeyForFederation(root: any) {
  return root;
}

export function filterInternalFieldsAndTypes(finalSchema: GraphQLSchema) {
  const removeEntitiesField = new FilterRootFields(
    (operation, fieldName) =>
      !(operation === 'Query' && (fieldName === '_entities' || fieldName === '_sdl')),
  );
  const removeEntityAndAny = new FilterTypes(
    type =>
      type.name !== '_Entity' &&
      type.name !== '_Any' &&
      type.name !== '_FieldSet' &&
      type.name !== '_Service' &&
      !type.name.startsWith('link__'),
  );
  const fakeSubschemaConfig = {
    schema: finalSchema,
    transforms: [removeEntitiesField, removeEntityAndAny],
  };
  for (const transform of fakeSubschemaConfig.transforms) {
    finalSchema = transform.transformSchema(finalSchema, fakeSubschemaConfig);
  }
  return finalSchema;
}
