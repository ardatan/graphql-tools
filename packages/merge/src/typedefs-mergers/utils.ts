import { Kind, ListTypeNode, NamedTypeNode, NonNullTypeNode, Source, TypeNode } from 'graphql';

export function isStringTypes(types: any): types is string {
  return typeof types === 'string';
}

export function isSourceTypes(types: any): types is Source {
  return types instanceof Source;
}

export function extractType(type: TypeNode): NamedTypeNode {
  let visitedType = type;
  while (visitedType.kind === Kind.LIST_TYPE || visitedType.kind === 'NonNullType') {
    visitedType = visitedType.type;
  }
  return visitedType;
}

export function isWrappingTypeNode(type: TypeNode): type is ListTypeNode | NonNullTypeNode {
  return type.kind !== Kind.NAMED_TYPE;
}

export function isListTypeNode(type: TypeNode): type is ListTypeNode {
  return type.kind === Kind.LIST_TYPE;
}

export function isNonNullTypeNode(type: TypeNode): type is NonNullTypeNode {
  return type.kind === Kind.NON_NULL_TYPE;
}

export function printTypeNode(type: TypeNode): string {
  if (isListTypeNode(type)) {
    return `[${printTypeNode(type.type)}]`;
  }

  if (isNonNullTypeNode(type)) {
    return `${printTypeNode(type.type)}!`;
  }

  return type.name.value;
}

export enum CompareVal {
  A_SMALLER_THAN_B = -1,
  A_EQUALS_B = 0,
  A_GREATER_THAN_B = 1,
}
export type CompareFn<T> = (a: T | undefined, b: T | undefined) => -1 | 0 | 1;

export function defaultStringComparator(a: string | undefined, b: string | undefined): CompareVal {
  if (a == null && b == null) {
    return CompareVal.A_EQUALS_B;
  }
  if (a == null) {
    return CompareVal.A_SMALLER_THAN_B;
  }
  if (b == null) {
    return CompareVal.A_GREATER_THAN_B;
  }
  if (a < b) return CompareVal.A_SMALLER_THAN_B;
  if (a > b) return CompareVal.A_GREATER_THAN_B;
  return CompareVal.A_EQUALS_B;
}
