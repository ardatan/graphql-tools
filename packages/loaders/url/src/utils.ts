import { memoize1 } from '@graphql-tools/utils';
import { OperationDefinitionNode } from 'graphql';
import type { Readable } from 'stream';

export function isBlob(obj: any): obj is Blob {
  return typeof obj.arrayBuffer === 'function';
}

interface GraphQLUpload {
  filename: string;
  mimetype: string;
  createReadStream: () => Readable;
}

export function isGraphQLUpload(upload: any): upload is GraphQLUpload {
  return typeof upload.createReadStream === 'function';
}

export function isPromiseLike(obj: any): obj is PromiseLike<any> {
  return typeof obj.then === 'function';
}

export const isLiveQueryOperationDefinitionNode = memoize1(function isLiveQueryOperationDefinitionNode(
  node: OperationDefinitionNode
) {
  return node.directives?.some(directive => directive.name.value === 'live');
});

export enum LEGACY_WS {
  CONNECTION_INIT = 'connection_init',
  CONNECTION_ACK = 'connection_ack',
  CONNECTION_ERROR = 'connection_error',
  CONNECTION_KEEP_ALIVE = 'ka',
  START = 'start',
  STOP = 'stop',
  CONNECTION_TERMINATE = 'connection_terminate',
  DATA = 'data',
  ERROR = 'error',
  COMPLETE = 'complete',
}
