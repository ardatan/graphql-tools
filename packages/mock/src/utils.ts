import { getRootTypeNames } from '@graphql-tools/utils';
import { GraphQLObjectType, GraphQLSchema } from 'graphql';
import { Ref, KeyTypeConstraints } from './types.js';

export function uuidv4() {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
    const r = (Math.random() * 16) | 0;
    // eslint-disable-next-line eqeqeq
    const v = c == 'x' ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}

export const randomListLength = () => {
  // Mocking has always returned list of length 2 by default
  // return 1 + Math.round(Math.random() * 10)
  return 2;
};

export const takeRandom = <T>(arr: T[]) => arr[Math.floor(Math.random() * arr.length)];

export function makeRef<KeyT extends KeyTypeConstraints = string>(typeName: string, key: KeyT): Ref<KeyT> {
  return { $ref: { key, typeName } };
}

export function isObject(thing: any) {
  return thing === Object(thing) && !Array.isArray(thing);
}

export function copyOwnPropsIfNotPresent(target: Record<string, any>, source: Record<string, any>) {
  for (const prop of Object.getOwnPropertyNames(source)) {
    if (!Object.getOwnPropertyDescriptor(target, prop)) {
      const propertyDescriptor = Object.getOwnPropertyDescriptor(source, prop);
      Object.defineProperty(target, prop, propertyDescriptor == null ? {} : propertyDescriptor);
    }
  }
}

export function copyOwnProps(target: Record<string, any>, ...sources: Array<Record<string, any>>) {
  for (const source of sources) {
    let chain = source;
    while (chain != null) {
      copyOwnPropsIfNotPresent(target, chain);
      chain = Object.getPrototypeOf(chain);
    }
  }
  return target;
}

export const isRootType = (type: GraphQLObjectType, schema: GraphQLSchema) => {
  const rootTypeNames = getRootTypeNames(schema);

  return rootTypeNames.has(type.name);
};
