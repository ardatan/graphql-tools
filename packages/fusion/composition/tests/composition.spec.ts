import { buildSchema } from 'graphql';
import {
  composeSubgraphs,
  createRenameFieldTransform,
  createRenameTypeTransform,
} from '@graphql-tools/fusion-composition';
import { printSchemaWithDirectives } from '@graphql-tools/utils';

describe('Composition', () => {
  const aSchema = buildSchema(/* GraphQL */ `
    type Query {
      myFoo: Foo!
    }

    type Foo {
      id: ID!
    }
  `);

  const bSchema = buildSchema(/* GraphQL */ `
    type Query {
      foo(id: ID!): Foo!
    }

    type Foo {
      id: ID!
      bar: String!
    }
  `);
  it('composes basic schemas', () => {
    const composedSchema = composeSubgraphs([
      {
        name: 'A',
        schema: aSchema,
      },
      {
        name: 'B',
        schema: bSchema,
      },
    ]);

    expect(printSchemaWithDirectives(composedSchema)).toMatchSnapshot();
  });
  it('composes with transforms on types', () => {
    const prefixTransform = createRenameTypeTransform(
      (type, subgraphConfig) => `${subgraphConfig.name}_${type.name}`,
    );
    const composedSchema = composeSubgraphs([
      {
        name: 'A',
        schema: aSchema,
        transforms: [prefixTransform],
      },
      {
        name: 'B',
        schema: bSchema,
        transforms: [prefixTransform],
      },
    ]);

    expect(printSchemaWithDirectives(composedSchema)).toMatchSnapshot();
  });
  it('composes with transforms on fields', () => {
    const prefixTransform = createRenameFieldTransform(
      (_field, fieldName, _typeName, subgraphConfig) => `${subgraphConfig.name}_${fieldName}`,
    );
    const composedSchema = composeSubgraphs([
      {
        name: 'A',
        schema: aSchema,
        transforms: [prefixTransform],
      },
      {
        name: 'B',
        schema: bSchema,
        transforms: [prefixTransform],
      },
    ]);

    expect(printSchemaWithDirectives(composedSchema)).toMatchSnapshot();
  });
});
