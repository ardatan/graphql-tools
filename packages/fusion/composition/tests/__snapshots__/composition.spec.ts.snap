// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Composition composes basic schemas 1`] = `
"schema {
  query: Query
}

type Query {
  myFoo: Foo! @resolver(subgraph: "A", operation: "query myFoo { myFoo }")
  foo(id: ID!): Foo! @resolver(subgraph: "B", operation: "query foo($id: ID!) { foo(id: $id) }")
}

type Foo @source(subgraph: "B", name: "Foo") {
  id: ID! @source(subgraph: "B", name: "id")
  bar: String! @source(subgraph: "B", name: "bar")
}"
`;

exports[`Composition composes with transforms on fields 1`] = `
"schema {
  query: Query
}

type Query {
  A_myFoo: Foo! @resolver(subgraph: "A", operation: "query myFoo { myFoo }")
  B_foo(id: ID!): Foo! @resolver(subgraph: "B", operation: "query foo($id: ID!) { foo(id: $id) }")
}

type Foo @source(subgraph: "B", name: "Foo") {
  A_id: ID! @source(subgraph: "A", name: "id")
  B_id: ID! @source(subgraph: "B", name: "id")
  B_bar: String! @source(subgraph: "B", name: "bar")
}"
`;

exports[`Composition composes with transforms on types 1`] = `
"schema {
  query: B_Query
}

type A_Query {
  myFoo: A_Foo! @resolver(subgraph: "A", operation: "query myFoo { myFoo }")
}

type A_Foo @source(subgraph: "A", name: "Foo") {
  id: ID! @source(subgraph: "A", name: "id")
}

type B_Query {
  foo(id: ID!): B_Foo! @resolver(subgraph: "B", operation: "query foo($id: ID!) { foo(id: $id) }")
}

type B_Foo @source(subgraph: "B", name: "Foo") {
  id: ID! @source(subgraph: "B", name: "id")
  bar: String! @source(subgraph: "B", name: "bar")
}"
`;
